server/src/main/java/org/opensearch/index/translog/RemoteFsTranslog.java

```diff
@@ -198,7 +202,17 @@ private boolean prepareAndUpload(Long primaryTerm, Long generation) throws IOExc
     }
 
     private boolean upload(Long primaryTerm, Long generation) throws IOException {
-        logger.trace("uploading translog for {} {} ", primaryTerm, generation);
+        // During primary relocation (primary-primary peer recovery), both the old and the new primary have engine
+        // created with the RemoteFsTranslog. Both primaries are equipped to upload the translogs. The primary mode check
+        // below ensures that the real primary only is uploading. Before the primary mode is set as true for the new
+        // primary, the engine is reset to InternalEngine which also initialises the RemoteFsTranslog which in turns
+        // downloads all the translogs from remote store and does a flush before the relocation finishes.
+        if (primaryModeSupplier.getAsBoolean() == false) {
+            logger.trace("skipped uploading translog for {} {}", primaryTerm, generation);
```

Comment:
Can we think about a No-op version of RemoteFsTranslog that switches atomically on tracker updates, rather than passing around primary mode here.

Reply:
Evaluated the approach you suggested. This is a bit tricky as currently the translog is supplied through translog factory and then the Translog implementation gets initialised only when the engine is created or reset. For the new primary during relocation, the engine is reset before the primary mode is changed from false into true and the primary mode change on older primary (during relocation) does not reset or recreates the engine. Although this is possible, this seems to be make this simple check to a bit complicated. Currently, changes in primaryMode has not listeners or observers. Will try this again after clearing of some high ROI items for remote-backed store items. Thanks for bringing this up.